### ISSUE TITLE
sample_posterior_predictive fails on dicts when samples not set

### ISSUE THREAD

**Issue by michaelosthege** [pymc-devs/pymc#3840]
## Description of your problem
https://github.com/pymc-devs/pymc3/blob/363afc80b0af4e873427420f10651cd86cc6263d/pymc3/sampling.py#L1523-L1525

This line breaks when `trace` is a list of dicts:
https://github.com/pymc-devs/pymc3/blob/master/pymc3/sampling.py#L1571

```
c:\users\osthege\repos\pymc3\pymc3\sampling.py in sample_posterior_predictive(trace, samples, model, vars, var_names, size, keep_size, random_seed, progressbar)
   1569 
   1570     if samples is None:
-> 1571         samples = sum(len(v) for v in trace._straces.values())
   1572 
   1573     if samples < len_trace * nchain:

AttributeError: 'list' object has no attribute '_straces'
```

Either `samples` should default to `len_trace*nchains`, or we should specify that `samples` must be set unless `trace` is a `MultiTrace`.

## Versions and main components
* PyMC3 Version: master

**Comment by rpgoldman** [comment]
@michaelosthege Do you have a quick MWE for this?  Or if not, would you mind testing on `fast_sample_posterior_predictive`?  I'd like to know if that has the same failure mode.

This is kind of an odd issue where it's not clear what arguments these functions do and do not accept.  And also, I have never used `find_MAP()`, so I didn't have a good test case.

**Comment by michaelosthege** [comment]
I didn't want to predict from MAP, but from a single posterior sample (a `dict`) I had tinkered with.

`fast_sample_posterior_predictive` works. Setting the `samples` kwarg also works, but that's not obvious..

```
with pymc3.Model() as pmodel:
    n = pymc3.Normal('n')
    trace = pymc3.sample()

with pmodel:
    d = pymc3.Deterministic('d', n * 4)
```

```
%%time
with pmodel:
    pp = pymc3.fast_sample_posterior_predictive(
        [trace[15]],
        var_names=['d']
    )
```

```
%%time
with pmodel:
    pp = pymc3.sample_posterior_predictive(
        [trace[15]],
        var_names=['d']
    )
```
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
...\pymc3\pymc3\sampling.py in sample_posterior_predictive(trace, samples, model, vars, var_names, size, keep_size, random_seed, progressbar)
   1569 
   1570     if samples is None:
-> 1571         samples = sum(len(v) for v in trace._straces.values())
   1572 
   1573     if samples < len_trace * nchain:

AttributeError: 'list' object has no attribute '_straces'
```

**Comment by rpgoldman** [comment]
I'm working on a fix for this now.  Should be done momentarily.

**Comment by rpgoldman** [comment]
Fixed by #3841
